File: 042.Trapping Rain Water/Trapping Rain Water.py
# 方法一，我们观察其中的一个位置，单独考虑这个位置的容量是多少？毫无疑问，我们只要知道，这个位置左边最高的那个边，和右边最高的那个边，两者取小的，然后再减去本身的大小，那么结果就是这个位置的容量。所以，
# （1）从左向右进行扫描，获取每个位置的左边最高的边。
# （2）从右向左进行扫描，获取每个位置的右边最高的边。
# （3）再遍历一边，计算出每个位置的容量，累加，即结果。

class Solution:
    def trap(self, height):
 
        if not height: return 0
        n, res = len(height), 0
        left_max, right_max = [0] * n, [0] * n
 
        left_max[0] = height[0]
        for i in range(1, n):  # 从左向右扫描一遍，求出每个位置左边最高的边
            left_max[i] = max(height[i], left_max[i - 1])
 
        right_max[n - 1] = height[n - 1]
        for i in range(n-2, -1, -1):  # 从右向左扫描一遍，求出每个位置右边最高的边
            right_max[i] = max(height[i], right_max[i + 1])
 
        for i in range(1, n-1):  # 扫描每一个位置，用当前位置左右最短的边，作为长度，并减去当前位置的值，就是当前位置的容量
            res += min(left_max[i], right_max[i]) - height[i]
        return res
 
 
if __name__ == '__main__':
    height = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
    solu = Solution()
    print(solu.trap(height))
# 方法二， 使用双指针法，在左右两端分别设置指针left, right，并且用 left_max， right_max 记录走过之后的最长最高的边。例如，left_max 就是指针left 左边最长的线，right_max 同理。
# （1）如果 height[left] < height[right]，就是左指针指向的数字 < 右指针指向的数字（现在操作左指针）,  说明此时，左边已经走过的位置都小于 height[right]，通俗的讲，现在left指针位置的左右两边桶的短板，一定存在于左边，又因为 left_max记录了当前left位置的左边最高的那个边，所以当前left位置的容量就是只需考虑left_max ， height[left] 这个两个值的大小即可得出。

# 问题：为什么说左边已经走过的位置都小于 height[right]？因为两个指针的移动条件是，那个指针小就移动那一个。

# （2）如果height[left] >= height[right]，同理（1）。直到两个指针相遇。

class Solution:
    def trap(self, height):
 
        if not height: return 0
        left_max = right_max = res = 0
        left, right = 0, len(height) - 1
 
        while left < right:
            if height[left] < height[right]:  # 左指针操作
                if height[left] < left_max:
                    res += left_max - height[left]
                else:
                    left_max = height[left]
                left += 1  # 移动左指针
            else:
                if height[right] < right_max:  # 右指针操作
                    res += right_max - height[right]
                else:
                    right_max = height[right]
                right -= 1  # 移动右指针
        return res
 
 
if __name__ == '__main__':
    height = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]
    solu = Solution()
    print(solu.trap(height))

File: 042.Trapping Rain Water/Trapping Rain Water.java
    //双指针 不使用堆栈
    public int trap3(int[] height) {
        int length = height.length;
        if(length<=2){
            return 0;
        }
        
        //获得可以盛水的区间
        int startIndex = 0;
        while(startIndex<length-1 && height[startIndex]<=height[startIndex+1]){
            startIndex++;
        }
        
        int result = 0;
        
        //这里思路类似于上一段代码，只是可以填水的区间将值改为填水后的值
        int index = startIndex;
        while(++index < length){
            int currentHeight = height[index];
            if(currentHeight > height[startIndex]){
                for(int i = index-1 ; i > startIndex ; i--){
                    result += (height[startIndex] - height[i]);
                }
                startIndex = index;
            }else{
                for(int i = index ; i>0 && height[i] > height[i-1] ; i--){
                    result += (height[i] - height[i-1]);
                    height[i-1] = height[i];
                }
            }
        }
        return result;
    }
//思路二：动态编程 先遍历获得区间左右最大值再计算
// 思路一中，每一次对一个区间都要遍历整个数组才能获得左右最大值。
// 但是其实，从左往右遍历一次数组，可以获得各个区间的leftMax。
// 同理，从右往左遍历可以获得各个区间的rightMax。将这两个值都存在数组中，并对数组进行遍历，计算各个区间的盛水高度。时间复杂度为O(n),代码如下：
    public int trap3(int[] height){
        int length = height.length;
        //leftMax数组
        int[] left = new int[length];
        //rightMax数组
        int[] right = new int[length];
        
        int leftMax = 0;
        int rightMax = 0;
        for(int i = 0 ; i<length ; i++){
            leftMax = left[i] = Math.max(leftMax,    height[i]);
            rightMax = right[length-i-1] = Math.max(rightMax, height[length-i-1]);
        }
        int result = 0;
        for(int j = 0 ; j<length ; j++){
            result += Math.min(left[j], right[j]) - height[j];
        }
        return result;
    }
//思路三：堆栈的聪明使用
//在这里，堆栈允许我们渐进的通过横向分割而非之前传统的纵向分割的方式来累加计算盛水量。一旦当前区间的高度超过栈顶的元素，
//就代表栈顶元素有一个右边界。鉴于栈中的元素都是递减的，所以如果存在一个比栈顶元素大的栈中元素，则一定可以确定该区间的盛水量。代码如下：
    public int trap4(int[] height){
        int length = height.length;
        int result = 0, current = 0;
        
        Stack<Integer> s = new Stack<Integer>();
        
        while(current < length){
            while(!s.isEmpty() && height[current] > height[s.peek()]){
                int top = s.pop();
                if(s.isEmpty()){
                    break;
                }
                //获得两个节点之间的宽度
                int distance = current - s.peek() - 1;
                int tempHeight = Math.min(height[current], height[s.peek()]) - height[top];
                result += tempHeight  * distance;
            }
            s.push(current++);
        }
        return result;
    }
//思路四：双指针的进阶
    public int trap5(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int result = 0;
        int leftMax=0, rightMax=0;
        while(left < right){
            if(height[left] < height[right]){
                leftMax = Math.max(height[left], leftMax);
                result += leftMax - height[left];
                left++;
            }else{
                rightMax = Math.max(height[right], rightMax);
                result += rightMax - height[right];
                right--;
            }
        }
        return result;
    }